castai/resource_castai_cluster.go

```diff
@@ -217,7 +384,107 @@ func waitForClusterToReachStatusFunc(ctx context.Context, client *sdk.ClientWith
+	return clusterLimitsCPU
+}
+
+func expandNodeDownscalerPolicy(params interface{}) *sdk.NodeDownscaler {
+
+	var nodeDownscalerPolicy sdk.NodeDownscaler
+	var nodeDownscalerPolicyEnabled sdk.NodeDownscalerEmptyNodes
+
+	for _, val := range params.([]interface{}) {
+		ndData := val.(map[string]interface{})
+		nodeDownscalerEnabled := ndData[PolicyFieldEnabled].(bool)
+		nodeDownscalerPolicyEnabled = sdk.NodeDownscalerEmptyNodes{
+			Enabled: &nodeDownscalerEnabled,
+		}
+	}
+
+	nodeDownscalerPolicy = sdk.NodeDownscaler{
+		EmptyNodes: &nodeDownscalerPolicyEnabled,
+	}
+
+	return &nodeDownscalerPolicy
+}
+
+func expandHeadroomPolicy(params interface{}) sdk.Headroom {
+	var headroomPercentage sdk.Headroom
+	for _, val := range params.([]interface{}) {
+		hpData := val.(map[string]interface{})
+		headroomPercentage = sdk.Headroom{
+			CpuPercentage:    hpData[PolicyFieldUnschedulablePodsHeadroomCPUp].(int),
+			MemoryPercentage: hpData[PolicyFieldUnschedulablePodsHeadroomRAMp].(int),
+		}
+	}
+	return headroomPercentage
+}
+
+func updatePolicies(ctx context.Context, client *sdk.ClientWithResponses, clusterID string, pc map[string]interface{}) diag.Diagnostics {
+	var clusterLimits sdk.ClusterLimitsPolicy
+	for _, val := range pc[PolicyFieldClusterLimits].([]interface{}) {
+		limitData := val.(map[string]interface{})
+
+		clusterLimits = sdk.ClusterLimitsPolicy{
+			Enabled: limitData[PolicyFieldEnabled].(bool),
+			Cpu:     expandCPUlimitPolicy(limitData[PolicyFieldClusterLimitsCPU].(interface{})),
+		}
+	}
+	log.Printf("[DEBUG] Reading Policies `PolicyFieldClusterLimits` parameterEnabled=#{clusterLimits}")
+
+	var nodeDownscalerPolicy sdk.NodeDownscaler
+	for _, val := range pc[PolicyFieldNodeDownscaler].([]interface{}) {
+		ndData := val.(map[string]interface{})
+		nodeDownscalerPolicy = *expandNodeDownscalerPolicy(ndData[PolicyFieldNodeDownscalerEmptyNodes].(interface{}))
+	}
+	log.Printf("[DEBUG] Reading Policies `PolicyFieldNodeDownscaler` parameterEnabled=#{&nodeDownscalerPolicy.EmptyNodes.Enabled}")
+
+	var spotInstancesPolicy sdk.SpotInstances
+	for _, val := range pc[PolicyFieldSpotInstances].([]interface{}) {
+		siData := val.(map[string]interface{})
+
+		spotInstancesPolicy = sdk.SpotInstances{
+			Enabled: siData[PolicyFieldEnabled].(bool),
+			Clouds:  convertStringArr(siData[PolicyFieldSpotInstancesClouds].([]interface{})),
+		}
+	}
+	log.Printf("[DEBUG] Reading Policies `PolicyFieldSpotInstances` parameterEnabled=#{spotInstancesPolicy}")
+
+	var unschedulablePodsPolicy sdk.UnschedulablePodsPolicy
+	for _, val := range pc[PolicyFieldUnschedulablePods].([]interface{}) {
+		upData := val.(map[string]interface{})
+
+		unschedulablePodsPolicy = sdk.UnschedulablePodsPolicy{
+			Enabled:  upData[PolicyFieldEnabled].(bool),
+			Headroom: expandHeadroomPolicy(upData[PolicyFieldUnschedulablePodsHeadroom].(interface{})),
+		}
+	}
+	log.Printf("[DEBUG] Reading Policies `PolicyFieldUnschedulablePods` parameterEnabled=#{unschedulablePodsPolicy}")
```

Comment:
drop all log.printf above and print entire autoscalerConfig once at line 483 (under autoscalerConfig assignment